### ISSUE TITLE
Add work arounds for first child like CSS

### ISSUE THREAD

**Issue by madou** [compiled/css-in-js#31]
Base library will always add a style tag before the styled element, e.g;

```
<>
  <style></style> // <-- this one
  <div></div>
</>
```

When a consumer is targetting `:first-child` and the like - perhaps we can transform it to `:first-child:not(style)` as a out of the box work around?

**Comment by madou** [comment]
This works for getting around `:first-child` problems: https://codesandbox.io/s/ssr-first-child-mitigation-7bmsj?file=/src/App.js

Now it's just a matter of figuring out `nth-child` - which tbh is infinitely more hard because each child may or may not be a compiled component.

**Comment by madou** [comment]
`:nth-child` workaround could be just a suggestion to code in data attributes.

```
:nth-child(2n) {
  background-color: gray;
}

<div />
<div /> // highlight
<div />
<div /> // highlight
```

```
[data-highlight] {
  background-color: gray;
}

<div />
<div data-highlight /> // highlight
<div />
<div data-highlight /> // highlight
```

**Comment by jesstelford** [comment]
It's possible to work around it in emotion too: https://github.com/emotion-js/emotion/issues/1178#issuecomment-469098353

**Comment by madou** [comment]
yep but you need to opt into the SSR extraction stuff - instead of that we'll be extracting to CSS files as the "advanced" setup

this is to solve the out of the box SSR problems where it compiles your css into your javascript but results in an inline style element

**Comment by Andarist** [comment]
All sorts of patterns are problematic with style tags rendered throughout the tree (consider for example `*+*` and targeting element deeper in the tree rather than only direct descendants), `:not` might also come handy in rewriting input selectors.

**Comment by madou** [comment]
oh that's such a good point!

**Comment by madou** [comment]
love your work @TxHawks, thanks!

like you said everything has its trade-offs, we will have to keep this in mind. another alternative i can think of that wouldn't need a JS solution would be to keep track if there are any nested selectors and always render them, resulting in some duplications

so for example:

```
<style>
  .a { color: hotpink; }
  @media screen and (min-width: 600px) { .b { color: red; } }
</style>
<div class="a b">Reliably hotpink until 600px and then red</div>
<!-- ... -->

<style>
  .c { color: green; }
  @media screen and (min-width: 600px) { .b { color: red; } }
</style>
<div class="b c">Reliably green undtil 600px and then red</div>
```

which of course is betting that duplication will be, on average, smaller than the javascript boilerplate

but it's something we need to keep in mind regardless, thanks dude!

**Comment by TxHawks** [comment]
Thank @madou 

Duplication would indeed work in this naïve example, but in other cases, could also cause unexpected inconsistencies, because it still alters the cascade (definitions bubble up - `.b` in the second style block also affects the first div. See example [here](https://codepen.io/veganarchist/pen/PozYENq?editors=1000)).

This would be fine with the previous example, but would fail in this one, for instance (you'd also have to apply duplication to pseudo selectors, not just nested ones:
```html
<style>
.a { color: red; }
.b:focus { color: blue; }
.c:active { color: purple; }
</style>
<button class="a b c">I'll be unexpectedly blue when active because of the second style block</button>

<!-- ... -->

<style>
.a { color: red; }
.b:focus { color: blue; }
.d:active { color: hotpink; }
</style>
<button class="a b d"> * * * </button>
```

Progressively rendering `style` tags into the markup has a ton of edge cases in atomic css and creates a whole lot of extra complexity.

**Comment by TxHawks** [comment]
Another point is that size-wise, duplication doesn't save the JS boilerplate, it just defers it to a later time. You'd need it anyway to move the content of the different `style` blocks into the one in `head` when JS executes.

**Comment by Andarist** [comment]
Q: is it any easier to deal with this stuff when creating just a single extracted stylesheet? My guess is not rly - but i havent thought about ot extensively. Definitely it couldnt rely on atomic classes for pseudos and would have to create „per component” ones, right?

**Comment by TxHawks** [comment]
It's a lot more straightforward with a single stylesheet (extracted or in `head`). What you'd basically do is figure out the order ruleset types (e.g., lvfha, media queries, etc.) should be in, and have corresponding "buckets" in the stylesheet to which you render rulesets by type (with atomic classes the order inside a bucket is meaningless). 

In Fela, we create a `style` element for each media query, taking advantage of the fact you can specify a `media` attribute on `style` elements, and then sort for lvfha inside them.

It's a lot simpler to reason about a single, predictable, source of truth, especially in a language that is order-dependent like CSS.

**Comment by madou** [comment]
Extracted single atomic sheet is what we are aiming for for the final destination in app, but for the intermediate state ensuring a workable 0 config story is something we must offer. 

When we look at trade offs I think there are the must haves and the nice to haves when it comes to streaming support (or in our case, 0 config support)

Must haves would be ensuring that:

- layout 
- colour 

Are stable. And don't change before/after JS executes.

If:

- pseudo interaction states (hover, focus) not being stable
- doubling up selectors from duplicate tags

Until JS executes wouldnt be the worst I think.  Are there other categories of edge cases you can think of?

**Comment by TxHawks** [comment]
Feature queries (`@supports` blocks) also come to mind in the same way as media queries, but I'm sure there are other edge cases I'm just not thinking about right now.

I'm also not at all certain that the JS boilerplate required for the inline `script` based solution will be larger than duplication, especially non-trivially sized applications.

As an aside, regardless of how the server-side application of styles will happen, for the client-side `style` element insertion, I'd recommend Fela's approach to style elements, as it enables a very clear separation of concerns (global, fonts, media query), and by so eases the task of sorting rulesets.

**Comment by madou** [comment]
Cheers will do 🙂

**Comment by mattrq** [comment]
### Possible approach:
Make sure there is a way of clearly targeting the injected inline styles and either exclude them or account for them. In the examples, I'm going to use `style`, but a more specific selector could be used.

From a practical perspective, we can ignore the usage of `*-of-type` as these would need to be targeted at `style` (unlikely and incorrect). It also means we can use `*-of-type` selectors.

#### [Example](https://codesandbox.io/p/sandbox/3hrxn5)
[<img width="556" alt="Screenshot 2024-11-22 at 2 33 08 AM" src="https://github.com/user-attachments/assets/71e0faa5-7e4e-4c71-ab03-c4a67022d2a6">
](https://codesandbox.io/p/sandbox/3hrxn5)

#### The following map works and is safe once JS runs:
 * `:first-child` => `:not(style):nth-of-type(1)`
 * `:nth-child(2)` => `:not(style):nth-of-type(2)`
 * `:last-child` => `:not(style):nth-last-of-type(1)`
 * `:nth-last-child(2)` => `:not(style):nth-last-of-type(2)`
 * `* + *` => `:not(style) + :not(style),:not(style) + style + :not(style)` - _We only need to allow for one style to be added inbetween_

**Comment by mattrq** [comment]
FYI @itsdouges; [see above](https://github.com/atlassian-labs/compiled/issues/31#issuecomment-2491579207)

**Comment by itsdouges** [comment]
Hey mate. With extraction being the target for production there really isn't a need for inline styles anymore.

Instead we can leverage the style hoisting behavior introduced in React 19 during development and then keep having extraction when building for production.

And then the inline style elements become a thing of the past and we side step this whole issue. 

@kylorhall-atlassian if you're interested :-)
